airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,441 @@
+            and self.incremental_most_recent_value_seen is not None
+        ):
+            last_date_updated = max(current_stream_state["last_date_updated"], self.incremental_most_recent_value_seen)
+        else:
+            last_date_updated = self.incremental_most_recent_value_seen
+        if current_stream_state is not None and "last_values_seen" in current_stream_state:
+            last_values_seen = self.update_last_values_seen_from_current_stream_state(current_stream_state)
+        else:
+            last_values_seen = self.primary_key_last_value_seen
+
+        return {"last_values_seen": last_values_seen, "last_date_updated": last_date_updated}
+
+    def update_last_values_seen_from_current_stream_state(self, current_stream_state: MutableMapping[str, Any]):
+        stream_last_values = current_stream_state["last_values_seen"]
+        new_values = {}
+        for key in self.primary_key_column:
+            stream_value = stream_last_values[key] if key in stream_last_values else STARTING_PRIMARY_KEY_VALUE
+            non_stream_value = (
+                self.primary_key_last_value_seen[key] if key in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+            )
+            new_values[key] = max(stream_value, non_stream_value)
+        return new_values
+
+    # END STATE MANAGEMENT FUNCTIONS
+
+    # QUERY GENERATION FUNCTIONS
+    def has_composite_primary_key(self):
+        return len(self.primary_key_column) > 1
+
+    def generate_query(self, sync_mode, stream_slice):
+        if sync_mode == SyncMode.full_refresh:
+            return self.generate_full_refresh_query(stream_slice)
+        elif sync_mode == SyncMode.incremental:
+            return self.generate_incremental_query(stream_slice)
+        else:
+            raise Exception(f'Unsupported Sync Mode: {sync_mode}.  Please use either "full_refresh" or "incremental"')
+
+    # How does full refresh work?
+    # We first split things into stream slices.  Then, we start with filtering our SQL query by >= the start date of the stream slice
+    # and <= the end date of the stream slice.  We then sort by the primary key column (s).
+    #  As we fetch records, we start filtering by > the most recent primary key value we've seen.  Because primary keys
+    # can be composite, we need to handle the cases where the first column might have multiple rows for a single value (since the second
+    # or third row might have different values)
+    # Note:  Outside of the stream slice dates, we don't filter by dates in full refresh.  We instead rely on Netsuite's ids being incremental.
+    # example Query:
+    # SELECT TOP 10000 [every property is individually articulated, like  accountnumber, acquisitionsource, alcoholrecipienttype ...] FROM Customer
+    #    WHERE ((id > -10000)) AND lastmodifieddate >= to_timestamp('2025-01-01', 'YYYY-MM-DD') AND lastmodifieddate <= to_timestamp('2025-12-31', 'YYYY-MM-DD')
+    #    ORDER BY id ASC, lastmodifieddate ASC
```

Comment:
is there a downside to ordering by lastmodifieddate, id so it's the same order as incremental?

Reply:
To my understanding, this would make filtering by primary key impossible, since primary keys would no longer appear in incremental order.

Reply:
wouldn't they still appear in incremental order, within a given time range?

Reply:
They would.  I think the biggest issue is that several tables (50+) don't have an incremental column that we could order on.

So we could have full refresh use the incremental query when it exists, but when it doesn't, we'd still need to use this query.

Open to making that change if you think it's a good idea.  Should I?

Reply:
yeah - the difference I see is today there's query patterns:
1. full refresh no cursor: order by id
2. full refresh with a cursor: order by id, order by timestamp
3. incremental with cursor: order by timestamp, order by id

I'm suggesting we simplify to
1. full refresh no cursor: order by id
2. full refresh and incremental with cursor: order by timestamp, order by id